Maha's Overleaf conciseness recommendations (#245)
symbiotic-engineering/MDOcean: pubs/applied-ocean-research-model/sections/benchmarking.tex

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Validation and Benchmarking}
\label{sec:validation-benchmarking}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Dynamic Validation using WEC-Sim}\label{sec:dynamic-validation}
The popular time-domain hydrodynamic simulation software WEC-Sim \cite{ruehl_wec-simwec-sim_2024} is used to validate the dynamics module.
The WEC-Sim RM3 example is run with regular waves and with the device constrained to oscillate only in heave.
\ifdefined\DISSERTATION
    Notably, the RM3 geometry provided in WEC-Sim differs slightly from the published dimensions in the RM3 report, so for validation the dimensions input to MDOcean are adjusted to match WEC-Sim.
\fi

\ifdefined\DISSERTATION
    WEC-Sim runs utilize hydrodynamic coefficients obtained with the WAMIT BEM for dynamics, and control coefficients calculated with MDOcean for consistency.
    MDOcean is run with MEEM as usual, and separately also run with the WAMIT coefficients to distinguish differences caused by disparate hydrodynamic coefficients from those caused by the underlying dynamics.
\else

>     Two MDOcean configurations are compared against WEC-Sim: one using MEEM hydrodynamic coefficients (the default) and one using WAMIT BEM coefficients matching WEC-Sim's, to separate dynamics-model error from hydrodynamic-coefficient error.

Author: For both diss and paper:
```suggestion
    To separate dynamics modeling error from hydrodynamic coefficient error, we compare two MDOcean configurations against WEC-Sim: one using MEEM hydrodynamic coefficients (the default) and one using WAMIT BEM coefficients matching WEC-Sim's.
    WEC-Sim uses identical control coefficients to MDOcean for consistency.
```

    \label{fig:error-histogram}
\end{figure}

> Results reveal that the drag describing function and MEEM hydrodynamic coefficients have a minor effect assuming a 1-DOF system (9.7\% and 2.7\% error on the average power and maximum amplitude respectively) but a major effect on the 2-DOF system (38.2\% and 28.6\% respectively).

Author: bring this sentence back, but talk about only the 2DOF results instead of both 1 and 2 DOF

> The extremely low errors in the 2-DOF system enforcing the same hydrodynamic coefficients as WEC-Sim and with zero drag (0.2\% and 1.9\% in power and amplitude respectively) indicate that this is not an error in the 2-DOF dynamic model itself, but in the way that a 2-DOF model amplifies errors in drag and hydrodynamic coefficients due to the importance of the phase of motion between each DOF. 

Author: bring this sentence back

> These errors are deemed acceptable for the purposes of this study, since the goal is to demonstrate the value of simultaneously analyzing multiple disciplines and the ability to quickly evaluate a large number of design options.

Author: ```suggestion
These errors are deemed acceptable for the purposes of this study, since the goal is to demonstrate the value of simultaneously analyzing multiple disciplines and the ability to quickly evaluate many designs.
```

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Static Validation}

\ifdefined\DISSERTATION
    The MEEM hydrodynamic coefficients for the nominal RM3 geometry match WAMIT BEM results closely, as previously shown in Fig.~5 of \cite{mccabe_open-source_2024}, except at frequencies below 0.3~rad/s.

>     This is expected because the BEM solution assumes infinite depth, which requires a finite added mass at low frequency, whereas MEEM uses finite depth, where the added mass must grow logarithmically toward infinity at zero frequency \cite{mciver_added_1991}.

Author: ```suggestion
    This low-frequency error is expected because the BEM solution assumes infinite depth, which requires a finite added mass at low frequency, whereas MEEM uses finite depth, where the added mass must grow logarithmically toward infinity at zero frequency \cite{mciver_added_1991}.
```

    For the site under consideration, the lowest-frequency sea state containing any energy is 0.4~rad/s, so the discrepancy is not a concern.

>     After the publication of reference~\cite{mccabe_open-source_2024}, excitation phase was added to the MEEM implementation. It also matches WAMIT results well.

Author: bring this sentence back

\ifdefined\DISSERTATION
    The MEEM hydrodynamic coefficients for the nominal RM3 geometry match WAMIT BEM results closely, as previously shown in Fig.~5 of \cite{mccabe_open-source_2024}, except at frequencies below 0.3~rad/s.
    This is expected because the BEM solution assumes infinite depth, which requires a finite added mass at low frequency, whereas MEEM uses finite depth, where the added mass must grow logarithmically toward infinity at zero frequency \cite{mciver_added_1991}.
    For the site under consideration, the lowest-frequency sea state containing any energy is 0.4~rad/s, so the discrepancy is not a concern.
    After the publication of reference~\cite{mccabe_open-source_2024}, excitation phase was added to the MEEM implementation.
It also matches WAMIT results well.
\else
    The MEEM hydrodynamic coefficients for the nominal RM3 geometry match WAMIT BEM results closely for $\omega \geq 0.3$~rad/s (Fig.~5 of \cite{mccabe_open-source_2024}); the divergence at lower frequencies arises from differing finite-depth versus infinite-depth assumptions \cite{mciver_added_1991} and is not a concern for the site considered (minimum sea-state frequency 0.4~rad/s).
\fi

> The overall model is validated by comparing simulated structural forces, power, mass, and cost results to the nominal values in \cite{RM3}, as shown in \cref{tab:validation}.

Author: bring this line back bc otherwise the table isn't referenced.
```suggestion
\Cref{tab:validation} compares simulated forces, power, mass, and cost to nominal RM3 values \cite{RM3} to verify overall accuracy.
```

\fi

\ifdefined\DISSERTATION
    As with force, a tuned scale factor is also applied to the power when the geometry is set to the nominal RM3 geometry to obtain an equal average power to the reference model report (86~kW).
    Unlike the force, the power discrepancy is not attributed to a modeling error because the separate power validation against WEC-Sim tracks extremely well as described in \cref{sec:dynamic-validation}.
    Rather, the discrepancy is attributed to the fact that the reference model report uses a different dynamic model that is not described in detail, and does not document the hydrodynamic coefficients, water depth, drag coefficient, and other necessary input parameters that would be required to obtain a close match.
    A final tuned scale factor is applied to the structural mass to account for the fact that not all structural components are modeled, including spar stiffeners, damping plate circumferential stiffeners, and miscellaneous features like flanges, enclosures, and access mounts.
\else

>     A power scale factor is similarly applied to match the reference report's 86~kW average power; the discrepancy is attributed to the reference report's use of an undocumented dynamic model rather than to a modeling error in MDOcean (the WEC-Sim validation of \cref{sec:dynamic-validation} tracks closely).

Author: ```suggestion
    A power scale factor is similarly applied to match the reference report's 86~kW average power; the discrepancy is attributed to the reference report's use of an undocumented dynamic model rather than to a modeling error in MDOcean because the WEC-Sim validation of \cref{sec:dynamic-validation} tracks closely.
```

> A structural-mass scale factor accounts for unmodeled subcomponents (spar stiffeners, damping plate circumferential stiffeners, flanges, enclosures, access mounts).

Author: ```suggestion
A structural-mass scale factor accounts for unmodeled subcomponents (spar stiffeners, damping plate circumferential stiffeners, flanges, enclosures, and access mounts).
```

%\hl{Explain sources of error and rough uncertainty and the implications of what we can trust}

The scaling behavior of economic outputs against the number of WECs is validated by sweeping $N_{WEC}$ from 1 to 100 and comparing the simulated CAPEX, OPEX, and LCOE against the values in the reference model report \cite{RM3}.
The average percent errors are included in \cref{tab:validation}, with additional detail in \cref{fig:econ-nwec-validate} of \appendixname~\ref{sec:appendix-econ-validation}.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Runtime Benchmarking}
\label{sec:sim-runtime}
Benchmarking the runtime of the MDOcean simulation is important to verify it achieves the desired speed to facilitate rapid design optimization.
An initial speed requirement order of magnitude of 100~ms for all modules was set to enable a 100-iteration finite difference optimization with 12 design variables to complete in around two minutes.
Ultimately, each simulation run takes around \resultsAOR[simRuntime], solidly meeting the goal.

> The timings in this section are performed on an Ubuntu 20.04 LTS server with a 14-core Intel Core i9-10940X CPU (3.3~GHz base clock) and 256~GB of DDR4 RAM at 3200~MHz, running MATLAB R2024b.

Author: ```suggestion
Timings in this section utilize an Ubuntu 20.04 LTS server with a 14-core Intel Core i9-10940X CPU (3.3~GHz base clock) and 256~GB of DDR4 RAM at 3200~MHz, running MATLAB R2024b.
```

\ifdefined\DISSERTATION
    The biggest portion is dedicated to evaluating Bessel functions in the semi-analytical solution, another large portion to unpacking variables from the cell data structure, and a smaller portion to solving the imaginary modes of the dispersion relation and the linear matrix equation.
\fi
The simulation is an order of magnitude faster than the Capytaine BEM solver for similar convergence levels.

\begin{figure}[b!]
\centering
\includegraphics[width=0.5\linewidth]{figs/from-matlab/sim_runtime.pdf}
\caption{Bar chart showing simulation runtime breakdown between modules}\label{fig:runtime-modules}
\end{figure}

\begin{figure}[t!]
\centering
\includegraphics[width=0.5\linewidth]{figs/from-matlab/hydro_runtime_logscale.pdf}
\caption{Bar chart demonstrating the speed improvement of MDOcean's hydro module over baseline solver Capytaine}\label{fig:runtime-hydro}
\end{figure}

The dynamics and controls module takes the next longest (\resultsAOR[pctRuntimeDynamics], enlarged in \figureautorefname~\ref{fig:runtime-dynamics}), with contributions from force saturation, spar analysis, drag linearization, and evaluation of the motion transfer function.

> This represents a three-order-of-magnitude improvement over the equivalent regular-wave WEC-Sim simulation run in parallel.

Author: ```suggestion
In \resultsAOR[runtimeDynamics] (non-parallelized), MDOcean solves the full constrained optimal control problem, representing a three-order-of-magnitude improvement over the parallelized WEC-Sim runtime to evaluate a single fixed controller.
Both simulate all sea states as regular waves.
```

\ifdefined\DISSERTATION

>     Simplifying the dynamics to a single degree of freedom (DOF) achieves another order of magnitude speedup, although the optimization and benchmarking results presented here utilize the 2-DOF model.

Author: add sentence about how early sims had even faster runtimes that could be obtained with even more simplifying assumptions, as in a multi-fidelity optimization.

\fi

\begin{figure}[t!]
\centering
\includegraphics[width=0.5\linewidth]{figs/from-matlab/dynamics_runtime.pdf}
\caption{Bar chart demonstrating the speed improvement of MDOcean's dynamics module over baseline solver WEC-Sim}\label{fig:runtime-dynamics}
\end{figure}

The structures, geometry, and economics modules are not computationally expensive and together compose the remaining \resultsAOR[pctRuntimeOther]~of the total runtime.

% say that it's 210 sea states, that wecsim is parallelized across X cores, that MEEM is using N=M=K=10

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Discussion}
\ifdefined\DISSERTATION

>     \Cref{sec:validation-benchmarking} demonstrates that MDOcean achieves accuracies generally within single-digit-percent JPD-weighted annual average power under realistic conditions, despite worst-case per-sea-state errors that can be larger due to drag and inter-body phase sensitivity in the 2-DOF model %within 10\% of established models and benchmarks both at the individual subsystem level and at the system level.

Author: ```suggestion
    \Cref{sec:validation-benchmarking} demonstrates that MDOcean achieves accuracies generally within single-digit-percent JPD-weighted annual average power and maximum amplitude under realistic conditions, despite worst-case per-sea-state errors that can be larger due to drag and inter-body phase sensitivity%within 10\% of established models and benchmarks both at the individual subsystem level and at the system level.
```

    The model is quite accurate for inputs near the nominal RM3 design and therefore its ability to capture multidisciplinary interactions in a single software package while running 1-3 orders of magnitude faster than established software is considered a major success.
    Still, the model is not perfect, and although only minor tuning was performed, the model's development centers around the Reference Model 3 and there may be unidentified discrepancies for geometries far from the reference design.
    We emphasize that the model is not intended to replace detailed models for final design and analysis, but rather to enable rapid design space exploration and optimization in the early stages of design.
    Since the values without scale factors are of the correct order of magnitude and the trends are reasonable, the use of scale factors to resolve discrepancies in the force, power, and mass models is deemed acceptable for the purposes of this study.
\else

>     MDOcean reproduces the dynamics of an established time-domain solver within a few percent under matched modeling assumptions and within single-digit-percent JPD-weighted annual average power under realistic conditions, despite worst-case per-sea-state errors that can be larger due to drag and inter-body phase sensitivity in the 2-DOF model.

Author: ```suggestion
    MDOcean reproduces the dynamics of an established time-domain solver within 0.2\% under matched modeling assumptions and within single-digit-percent JPD-weighted annual average power under realistic conditions, despite worst-case per-sea-state errors that can be larger due to drag and inter-body phase sensitivity.
```

Three scale factors are used to tune force, power, and mass outputs across disciplines for consistency with the reference model report.

> The \resultsAOR[simRuntime] runtime is 1-3 orders of magnitude faster than established software, enabling multidisciplinary optimization workflows that would otherwise be prohibitive.

Author: ```suggestion
The \resultsAOR[simRuntime] full-simulation runtime is 1-3 orders of magnitude faster than established software, enabling multidisciplinary optimization workflows that would otherwise be prohibitive.
```
